### PR TITLE
MacOSX Support

### PR DESCRIPTION
**Pull request by kangaroo**
Alright, total brain dump pull request.  I had some time and made mac work.  Simple CLR programs run, more complicated stuff does not yet.

A few notable things:
- binary commit of config.h sucks, but its there on linux too, and all your configure.in dependencies are missing
- fixup precode is currently causing unaligned calls on OS X, haven't bothered looking into it yet
- some of the assembly refactoring is ugly, not sure what direction you want to go on for the future
- I nuked with abandon the old silverlight .bundle stuff
- release build doesn't work on mac yet
- linux builds work for me on ubuntu, haven't tested windows (buildbot be kind to me I hope)
- your cmake files seem to be auto generated for linux, but you don't have the generation script, thus the 1000+ file change

I'm not sure how much time I'll have to do serious refactoring if required, but lets discuss.  It might be preferable to squash this?  In the worst case this is a kick off point for others who want to do bring-up.

### CODE REVIEW

**Comment by kangaroo** [comment]
Seems the CI linux fails on your builder.  Any chance of getting a VERBOSE=1 build to figure out why?

**Comment by kangaroo** [comment]
Windows build is fixed.  Your linux build can't seem to find ar based on one of my changes. I'm on 64-bit trusty.  Can you please let me know what you're buildbot is verifying against?

**Comment by borgdylan** [comment]
Is enabling the build on 32 bit linux possible?

**Comment by sergiy-k** [comment]
@kangaroo This is awesome! Thank you so much for doing it! I’ll start reviewing it in the morning.

Regarding conig.h: the plan is to be able to generate it automatically. The one that we currently have in the source tree is a temporary thing that I created in almost automated way with some manual adjustments. :) I do have configure.in but it needs a number of fixes before I can check it in.

Once again, thank you for your contribution!

**Comment by sergiy-k** [comment]
@kangaroo ... and, by the way, about our lab verification. We are using Ubuntu 14.04 with llvm 3.5 and cmake 2.8.12 installed. Currently this is the only configuration that we are testing and supporting. Thank you.

File: CMakeLists.txt [hunk 1]
```diff
@@ -283,17 +292,25 @@ endif (WIN32)
...
+
+  if(CLR_CMAKE_PLATFORM_LINUX)
+    add_definitions(-D__LINUX__=1)
+    if(CMAKE_SYSTEM_PROCESSOR STREQUAL x86_64)
+      message("Detected Linux x86_64")
+      add_definitions(-DLINUX64)
+    endif(CMAKE_SYSTEM_PROCESSOR STREQUAL x86_64)
+  endif(CLR_CMAKE_PLATFORM_LINUX)
+  if(CLR_CMAKE_PLATFORM_DARWIN)
+    add_definitions(-D_XOPEN_SOURCE)
```
Comment (shahid-pk) [inline CMakeLists.txt]: i think here also should be a message like detected darwin.

File: src/pal/src/CMakeLists.txt [hunk 2]
```diff
@@ -128,14 +140,30 @@ set(SOURCES
...
   ${SOURCES}
+  ${PLATFORM_SOURCES}
 )
 
 target_link_libraries(CoreClrPal
-  pthread
-  rt
   dl
   m
 )
```
Comment (shahid-pk) [inline src/pal/src/CMakeLists.txt]: does this disable pthread and rt on windows ? i think they are required their.

**Comment by jkotas** [comment]
Could you please submit the change in the 1000+ autogenerated test CMakeFiles as separate PR, so that the rest is easier to deal with.

File: CMakeLists.txt (hunk 1 above)
Comment (kangaroo) [inline CMakeLists.txt]: Added in 716ae46

**Comment by kangaroo** [comment]
@sergiy-k Hrmm.. my linux VM is the same, so I'm not sure why the link is failing.

**Comment by kangaroo** [comment]
@jkotas cherry-picked that out into https://github.com/dotnet/coreclr/pull/112

**Comment by kangaroo** [comment]
Linux build issue was a typo in the gen-buildsys-clang.sh.  Running a test build now, and should be committing it shortly.

**Comment by kangaroo** [comment]
I believe 8222c67 addresses all raised concerns on this PR.

Anything else?

File: build.sh [hunk 3]
```diff
@@ -62,6 +62,21 @@ check_prereqs()
 
 build_coreclr()
 {
+    #determine nproc in a platform independent way
+    NPROCS=1
+    OS=`uname`
+
+    if [ $OS = "Linux" ]; then
+      NPROCS=`grep -c ^processor /proc/cpuinfo`
```
Comment (sergiy-k) [inline build.sh]: I think this has been already address in a slightly different way in https://github.com/dotnet/coreclr/pull/100

Comment (kangaroo) [inline build.sh]: I'll merge it back once thats merged to master.

Comment (jkotas) [inline build.sh]: #100 was merged to master. Could you please also squash the commits?

File: src/pal/src/CMakeLists.txt [hunk 4]
```diff
@@ -9,6 +9,17 @@ include_directories(include)
...
+    arch/i386/context.clang.s
+    arch/i386/dispatchexceptionwrapper.S
+    exception/machexception.cpp
+    exception/machmessage.cpp
+    locale/locale.cpp
+  )
+endif(CMAKE_SYSTEM_NAME STREQUAL Darwin)
+
 add_definitions(-DPLATFORM_UNIX=1)
 add_definitions(-D__LINUX__=1)
```
Comment (sergiy-k) [inline src/pal/src/CMakeLists.txt]: This (__LINUX__=1) should not be defined when compiling for Mac, right?

File: src/vm/amd64/unixasmhelpers.S [hunk 5]
```diff
@@ -27,7 +27,7 @@ LEAF_ENTRY PrecodeFixupThunk, _TEXT
         lea     METHODDESC_REGISTER,[rax+r11*8]
 
         // Tail call to prestub
-        jmp     ThePreStub
+        call C_FUNC(ThePreStub)
```
Comment (sergiy-k) [inline src/vm/amd64/unixasmhelpers.S]: This does not look right. The call instruction will make the stack aligned incorrectly, right? Could this be the root cause of the issue in ThePreStub that you and @janvorli were discussing earlier?

Comment (kangaroo) [inline src/vm/amd64/unixasmhelpers.S]: Whoops.  Fixed that while rebasing the nproc changes.  I'll include it in the squash @jkotas requested.

File: src/pal/src/CMakeLists.txt (hunk 4 above)
Comment (kangaroo) [inline src/pal/src/CMakeLists.txt]: Also fixed in the upcoming squash

**Comment by kangaroo** [comment]
Grrr.. my rebase discarded my changes somehow.  Going to need to redo that.

**Comment by sergiy-k** [comment]
@kangaroo You mean you lost everything? I'm so sorry to hear that... :(

**Comment by kangaroo** [comment]
@sergiy-k no, I have a local copy branch, I just had to redo the rebase again.  GitHub is horribly confused, but the PR is @ https://github.com/dotnet/coreclr/pull/117 now

**Comment by sergiy-k** [comment]
@kangaroo You scared me. Now I'm again in good mood. :)

**Comment by kangaroo** [comment]
@sergiy-k Always make a backup before messing around with rebasing and force pushing :)

**Comment by cnblogs-dudu** [comment]
Great! It works!
```
git clone -b osx --single-branch https://github.com/kangaroo/coreclr.git
./build.sh                                             
```

**Comment by kangaroo** [comment]
@cnblogs-dudu I can upload a corlib you can use when I get back from dinner of you like. Building it is a little onerous right now

**Comment by kangaroo** [comment]
Sample corlib available at:

https://www.dropbox.com/s/zvl5tsj6peh12km/mscorlib.dll?dl=0

**Comment by cnblogs-dudu** [comment]
@kangaroo 
Thanks! Get it.

**Pull request closed without merging**